pkg/controller/uniteddeployment/allocator.go

```diff
@@ -302,58 +308,197 @@ func (ac *elasticAllocator) validateAndCalculateMinMaxMap(replicas int32, nameTo
+		replicas = *ac.Spec.Replicas
+	}
+	minReplicasMap, maxReplicasMap, err := calculateRawMinMaxMap(replicas, ac.Spec.Topology.Subsets)
+	if err != nil {
+		return nil, err
+	}
+	var subsets []string
+	for _, subset := range ac.Spec.Topology.Subsets {
+		subsets = append(subsets, subset.Name)
+	}
+	nextReplicas := allocateByMinMaxMapAndReservation(replicas, minReplicasMap, maxReplicasMap, existingSubsets, subsets)
+	klog.V(4).InfoS("got UnitedDeployment next replicas", "unitedDeployment",
+		klog.KObj(ac.UnitedDeployment), "nextReplicas", nextReplicas)
+	return nextReplicas, nil
+}
+
+type reservationAllocatorSubsetRecord struct {
+	maxAlloc int32
+	replicas int32
+}
+
+// allocate allocates a subset to the 'to' replicas and returns how many replicas is really consumed.
+//   - param to: will scale up the subset to the value
+//   - param rest: rest replicas can be allocated
+func (ar *reservationAllocatorSubsetRecord) allocate(to, rest int32) (allocated int32) {
+	toAdd := min(to-ar.replicas, rest)
+	ar.replicas += toAdd
+	ar.maxAlloc -= toAdd
+	return max(0, min(0, ar.maxAlloc)+toAdd)
+}
+
+func exportReservationRecords(records map[string]*reservationAllocatorSubsetRecord) map[string]int32 {
+	var nextReplicas = make(map[string]int32, len(records))
+	for name, record := range records {
+		nextReplicas[name] = record.replicas
+	}
+	return nextReplicas
+}
+
+// The explanation of this allocation algorithm is as follows.
+// When X replicas are assigned to a normal subset, the total number of replicas will normally decrease by X;
+// whereas when assigned to an unschedulable subset (subsets where all reserved pods have just started are still considered as unschedulable),
+// the total number will only reduce by at most the number of non-reserved Pods in the subset.
+// During allocation, each subset is first assigned a number of replicas equal to minReplicas in sequence.
+// Then, each subset is allocated up to maxReplicas (schedulable) or its allocated number (unschedulable) in sequence again.
+func allocateByMinMaxMapAndReservation(replicas int32, minReplicasMap, maxReplicasMap map[string]int32,
+	existingSubsets map[string]*Subset, subsetNames []string) map[string]int32 {
+	// Prepare for the records
+	records := make(map[string]*reservationAllocatorSubsetRecord, len(subsetNames))
+	for _, name := range subsetNames {
+		if subset := existingSubsets[name]; subset == nil || subset.Allocatable() {
+			records[name] = &reservationAllocatorSubsetRecord{maxAlloc: math.MaxInt32}
+		} else {
+			records[name] = &reservationAllocatorSubsetRecord{maxAlloc: subset.Spec.Replicas - subset.Status.UnschedulableStatus.ReservedPodNum}
+		}
+	}
+	// Step 1: satisfy the minimum replicas of each subset firstly.
+	for _, name := range subsetNames {
+		replicas -= records[name].allocate(minReplicasMap[name], replicas)
+		if replicas <= 0 {
+			return exportReservationRecords(records)
+		}
+	}
+
+	// Step 2: satisfy the reserved replicas of each unschedulable subset
```

Comment:
the comment is not precise since it also allocate for schedulable subset, a possible alternative is to allocate the rest  replicas into subsets sequentially